### PR TITLE
Issue #78 - Products GraphQL + Search

### LINKED ISSUE #78: Products GraphQL + Search
### **📁 Files**:

* `app/graphql/types/product_type.py`
* `app/graphql/resolvers/product_resolver.py`
* `app/services/search_service.py`

### **📋 Tasks**:

* [ ] Full `ProductType` with relationships (`farmer`, `category`)
* [ ] Advanced search: `searchProducts(filters: ProductSearchInput)`
* [ ] Aggregations: products by category, farmers by region
* [ ] Performance tests for product search

### **✅ Test Criteria**:

```bash
# GraphQL supports complex multi-filter product searches
# Nested queries like farmer -> products -> category work correctly
# Search response time under 500ms for datasets with 1000+ products
```

---

### PR DESCRIPTION
**Pull request by ricardomazuera**
# Pull Request RevolutionaryFarmers

## 📝 Summary

This PR creates the product search functionality in the GraphQL API, enabling more flexible and efficient queries for products in the marketplace.

---

### 🪧 Related Issues

- Closes #78

---

### 🏁 Type of Change

Mark with an `x` all the checkboxes that apply (like `[x]`).

- [X] 📝 Documentation (updates to README, docs, or comments)
- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] 👌 Enhancement (non-breaking change which adds functionality)
- [X] 💥 Breaking change (fix or feature that would cause existing functionality to change)

---

### 🔄 Changes Made

- Added new GraphQL queries and resolvers for advanced product search.
- Created filtering and pagination for product listings.
- Created models to farmers 

---

### 🚀 Implementation Details

- Implemented new resolvers in `product_resolver.py` to support advanced search features.
- Updated the GraphQL schema and types to include new search parameters.
- Enhanced the search service to handle complex queries and return paginated results.
- Ensured all new endpoints use async/await and proper dependency injection.

---

### 🛠 Technical Notes

- All new code follows FastAPI and Strawberry GraphQL best practices.
- Type hints and Pydantic validation are used throughout.
- The search logic is modular and can be extended for future requirements.

---

### ✅ Tests Results

- Added and updated unit tests for the search service and GraphQL resolvers.
- Manual testing done via GraphQL Playground to ensure correct query responses.

---

### Test Coverage

- [x] ✅ Unit Tests
- [ ] ✅ Integration Tests
- [x] ✅ Manual Testing

---

### 📸 Evidence

- All tests pass successfully (see attached test output).
- Screenshots of GraphQL Playground queries and responses are included below.

#### Search
Screenshot 2025-07-06 at 12 24 39 PM

Screenshot 2025-07-06 at 12 24 58 PM

#### Coverage test
Screenshot 2025-07-06 at 12 25 27 PM

#### Lint
Screenshot 2025-07-06 at 1 06 57 PM

---

### 🔍 Testing Notes

- Tested edge cases such as empty search results, invalid parameters, and pagination limits.

---

### 🔜 Next Steps

- Monitor performance in staging and optimize queries if needed.
- Gather user feedback for further improvements to search functionality.

### CODE REVIEW

File: app/services/search_service.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    if not farmer:
+        return None
+    return FarmerType(
+        id=farmer.id,
+        name=farmer.name,
+        email=farmer.email,
+        phone=farmer.phone,
+        location=farmer.location,
+        verified=farmer.verified,
+    )
```
[the remark this answers is not in the document]
Comment (ricardomazuera) [reply app/services/search_service.py]: Done

File: app/graphql/resolvers/product_resolver.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                FarmerType(
+                    id=f.id,
+                    name=f.name,
+                    email=f.email,
+                    phone=f.phone,
+                    location=f.location,
+                    verified=f.verified,
+                )
+            )
+        return out
```
[the remark this answers is not in the document]
Comment (ricardomazuera) [reply app/graphql/resolvers/product_resolver.py]: Remove it

**Comment by ricardomazuera** [comment]
CodeRabbit finished. Please review it @aguilar1x. Thx

Screenshot 2025-07-06 at 2 37 48 PM

**Comment by aguilar1x** [comment]
gm @ricardomazuera, 

This issue is blocked until the remaining issues are completed, as each issue depends on the others.

I'll let you know as soon as it's unblocked. Thank you so much for wanting to contribute to Revolutionary Farmers.

File: app/graphql/resolvers/product_resolver.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+)
+from strawberry.types import Info
+from app.services.search_service import search_products_service
+
+
+@strawberry.type
+class ProductResolver:
+    @strawberry.field
+    async def search_products(
+        self, info: Info[Any, Any], filters: Optional[ProductSearchInput] = None
```
Comment (patrick91) [inline app/graphql/resolvers/product_resolver.py:19]: you can use `strawberry.Info` directly (and don't need to pass the arguments, they are optional)

```suggestion
        self, info: strawberry.Info, filters: ProductSearchInput | None = None
```

but if you create a context type, you can also do 

`strawberry.info[ContextType]` and then `info.context` will be typed 😊

File: app/graphql/resolvers/product_resolver.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    ProductType,
+    ProductCategoryType,
+    ProductSearchInput,
+)
+from strawberry.types import Info
+from app.services.search_service import search_products_service
+
+
+@strawberry.type
+class ProductResolver:
```
Comment (patrick91) [inline app/graphql/resolvers/product_resolver.py:16]: I usually calle these `ProductQuery` 😊

File: app/graphql/resolvers/product_resolver.py (hunk 3 above)
Comment (ricardomazuera) [reply app/graphql/resolvers/product_resolver.py:19]: Thanks for the comment, I learned something new 😅

File: app/graphql/resolvers/product_resolver.py (hunk 4 above)
Comment (ricardomazuera) [reply app/graphql/resolvers/product_resolver.py:16]: Yes, in fact I had put it that way, until I saw that they had [left documentation](https://github.com/Crypto-Jaguars/Revo-Backend/blob/main/app/graphql/schema.py#L22) to call resolver as `Resolver` instead of `Query`. That's why I decided to leave it as the maintainers indicated.

What do you think @aguilar1x, should we put it as @patrick91 says?
